src/azul/plugins/repository/canned/__init__.py

```diff
@@ -69,7 +71,104 @@
 log = logging.getLogger(__name__)
 
 
-class CannedSourceRef(SourceRef[SimpleSourceSpec, 'CannedSourceRef']):
+@attrs.frozen(kw_only=True)
+class CannedSourceSpec(SourceSpec):
+    url: str
```

Comment:
Maybe this could be a `furl` instance?

Reply:
> Maybe this could be a furl instance?

Unfortunately, no. The `RepositoryPlugin` has a `sources` property that returns a set of source specs, so `CannedSourceSpec` can't contain a `furl` object since it isn't hashable.